Initial work on Adam Bashforth methods

obiajulu opened the pull request:
First working build of Adam Bashforth methods integrated in local ODE.jl repo. I made a AdamBash module, which is mostly for testing purposes. It easily allowed me to include the solver 'ode_ab' in the test from test/runtests.jl  It passed the test. 

@mauro3 and @jiahao, it would be great to hear your thoughts and seek you advice for how I could best improve my coding practices, especially to suite existing Julia code best. I see this solver as more or less a warm up, and I think it's best to make the small, correctable mistakes here than later on!

obiajulu committed:
preliminary working adam bashforth

obiajulu committed:
progress on adam bashforth made. Passed initial in house tests

obiajulu committed:
Implemented Adam Bashforth using 'AdamBash' module, and included the solver 'ode_ab' in runtests.jl. The solver passed this test

mauro3 commented:
Thanks for this.  I have quite a few comments, as expected at this stage.

The main points are:

(1) The values of `f(t[i], y[i])`, etc., are known and should be stored instead of re-calculated. As `ode_ms` mentions, some sort of ring-buffer would be ideal, but for now I'd just go with a vector like for `y`.  

(2) Instead of hard-coding the individual methods, do it in a loop; again `ode_md` does this well.

Run the `interface-tests.jl` tests as well.

The others I annotated in the code.

Workflow-wise: when sharing branches with other people it is often useful to mark them as "yours".  I always start mine with `m3/`, e.g. https://github.com/mauro3/Sundials.jl/tree/m3/example-fix. So this one could be `ob/a-b`.

The question now is whether and how to take this further.  To make an informed decision, let's read up in H,N&W92, sections:  III.1 on implicit Adams Methods, III.7, III.5 to p.400.  I think irrespective of the decision, you could implement an implicit A-B method using a PECP scheme (eq. 1.9, 1.11).

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
-##Implementation of the Adam Steps for order 2, 3 and 4
+using Polynomials;
+using Compat;
+include("ODE.jl")
```
Include this file in `ODE.jl`, i.e. the other way round.  Then the two `using` statements above are not needed anymore (are they needed at all?)

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
+include("ODE.jl")
 
+##Implementation of the Adam Steps for order 2, 3 and 4 using recursive formualae from Hairer
 abstract Integrator
```
Unused type.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
-@inline function adam_step(::Union{AdamBashforthStep{2},Type{AdamBashforthStep{2}}}, f, y, t, stepsize)
+@inline function adam_step(::Union{AdamBashforth{2},Type{AdamBashforth{2}}}, f, y, t, stepsize, i)
     ynext = y[i] + stepsize/2*(3*f(t[i], y[i])
     - 1*f(t[i-1],y[i-1]))
```
Indent lines like this such that it starts at the column of the opening `(`

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
+immutable AdamBashforth{N} end
 
-@inline function adam_step(::Union{AdamBashforthStep{2},Type{AdamBashforthStep{2}}}, f, y, t, stepsize)
+@inline function adam_step(::Union{AdamBashforth{2},Type{AdamBashforth{2}}}, f, y, t, stepsize, i)
```
The union is not needed, with the change below just do `::AdamBashforth{2}`

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
+        yvals[i] = yint[i]
+    end
+    ##Use Adam Basforth method for subsequent steps
+    Integrator = AdamBashforth{order}
```
`integrator = AdamBashforth{order}()`, the `()` makes an instance, usually preferred. (and then with lower-case `i`).

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
+    yvals[1] = y0
+
+    ##Use Runge-Kunta for initial points necessary to base Adam Basforth off of, if initial values not given
+    tint, yint= ODE.ode_ms(F,y0, tspan[1:order], order)
```
Note that `mod_ms` is also a A-B method, maybe use `ode45`.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,58 @@
+        yvals[i] = adam_step(Integrator,F,yvals,tspan,h[i-1],i-1)
+    end
+    ##Return y values
+    tspan, yvals
```
Style: I'd write a explicit `return`.

obiajulu committed:
stores f(tspan[i], y[i]) values in array ydot[]

obiajulu commented:
@mauro3, thanks for all of the comments. I am currently editing my code. Now that I have taken a closer look at ode_ms, I realize that you have already written the AB explicit method and thought that it might make most sense to go ahead and adapt into an implementation of the implicit AB method with the PECP Scheme (as you mentioned). Just wanted to see if you agree that this is a better use of time.

mauro3 commented:
Someone else wrote it, maybe "git blame" can tell you.  But, yes, adapting that would be good.

obiajulu committed:
We switch from explicit to implicit adam methods

A pretty big commit. We change from a reimplementation of Adam Methods
found in the implementation of ode_ms, to implicit adam methods (see
Hairer III.1 on implicit Adam methods). We followed closely the coding
style of ode_ms. We switch from using the ODE module in adam_bashforth.jl
to including adam_bashforth.jl in ODE.jl. Further, we change from manually
coding each Adam step, to simply storing or calculating the necessary
coefficients.

obiajulu commented:
@mauro3, so I coded the implicit Adams method with the last commit. It seems to be passing both runtests.jl and interface-tests.jl.

mauro3 commented:
I wasted a bunch of time this morning to implement the circular buffer mentioned in that comment of `ode_ms`, maybe you want to use it:

``` julia
"""
Circular Buffer which can be pushed to and indexed from the end.  Only
lengths which are powers of 2 are allowed.

cb[end] gives the last pushed entry.
cb[end-1] gives the second last pushed entry.

Note these pitfalls:

- cb[i] should probably not be used
- cb[end-n] will just wrap around if n>length(cb)
"""
type CircBuffer{T} <: AbstractArray{T,1}
    len::Int
    len_pow::Int # log2(len)
    ptr::Int # [1,len]: index of last push!ed element
    buf::Vector{T}
    function CircBuffer(len::Integer)
        arr = Array(T,len)
        len_pow = convert(Int, log(2,len))
        new(len, len_pow, 1, arr)
    end
end
CircBuffer(T,len) = CircBuffer{T}(len)

Base.length(A::CircBuffer) = A.len
Base.endof(A::CircBuffer) = A.ptr

# Note rem is slow: https://github.com/JuliaLang/julia/issues/8188.   Thus this general implementation is a bit slow:
#@inbounds Base.getindex(A::CircBuffer, i::Int) = A.buf[rem1(i, A.len)]
# Instead using this (which only works for len a power of 2) improves speed 10x:
rem1_(x,m, mpow2) = x-((x-1)>>mpow2)*m
@inbounds Base.getindex(A::CircBuffer, i::Int) = A.buf[rem1_(i, A.len, A.len_pow)]

"Note that setindex! will not shift the circular buffer."
Base.setindex!(A::CircBuffer, v, i::Int) = A.buf[rem1_(i, A.len, A.len_pow)]=v
@inbounds function Base.push!(A::CircBuffer, v)
    A.ptr = A.ptr<A.len ? A.ptr+1 : 1
    A.buf[A.ptr] = v
    A
end
Base.writemime(io::IO, m::MIME"text/plain", A::CircBuffer) = writemime(io, m, [A[end-i] for i=0:length(A)-1])
```

Performance on Julia0.4 is on par with arrays for `getindex`, `push!` is about 5x slower than `setindex!` on an array.  I think this should not matter for this application.

``` julia
n = 10^8
cb = CircBuffer(Int, 16)
ar = zeros(Int, 0)
sizehint!(ar,n)

function fillit!(A, n)
    for i in 1:n
        push!(A,i)
    end
    nothing
end

function setit!(A, n)
    for i in 1:n
        A[i] = i
    end
    nothing
end

# JIT them
fillit!(cb,3)
fillit!(zeros(Int,3),3)
setit!(cb,3)
setit!(zeros(Int,3),3)

@time 1 # warmup @time
println("fillit!")
@time fillit!(cb,n)
@time fillit!(ar,n)

println("setit!")
@time setit!(cb,n)
@time setit!(ar,n)

function getit(A, n)
    out = 0
    for i in 1:n
        out += A[i]
    end
    out
end
getit(cb,3)
getit(zeros(Int,3),3)

println("getit")
@time getit(cb,n);
@time getit(ar,n);
```

gives

```
  0.000005 seconds (148 allocations: 10.151 KB)
fillit!
  0.445645 seconds (4 allocations: 160 bytes)
  0.467509 seconds (4 allocations: 160 bytes)
setit!
  0.153312 seconds (4 allocations: 160 bytes)
  0.083288 seconds (4 allocations: 160 bytes)
getit
  0.095685 seconds (5 allocations: 176 bytes)
  0.088339 seconds (5 allocations: 176 bytes)
```

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
+##order is set to 4 by default
+##Adam Bashforth is a fixed step, fixed order, multistep Implicit method
+function ode_imp_ab(F::Function,y0, tspan,order=4 ::Integer)
+    if (0 <= order <= 3)
```
Table 2.1 p.372 says that implicit adams has oder k+1, so this should be `(1<=order<=4)`.

obiajulu replied:
Great. Taken care of.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
+##Adam Bashforth is a fixed step, fixed order, multistep Implicit method
+function ode_imp_ab(F::Function,y0, tspan,order=4 ::Integer)
+    if (0 <= order <= 3)
+        coef = am_imp_coefficients3
```
How about calling those `b_imp` and `b_exp`?

obiajulu replied:
Good idea. I'll make the change in the next commit.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
     end
 
-    stepsize = diff(tspan)
+    h = diff(tspan)
```
In PR #49 standardizes the variable names to `t, dt, y, dy`.  Use that here too.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
-    stepsize = diff(tspan)
+    h = diff(tspan)
     y = Array(typeof(y0), length(tspan))
+    ydot = similar(y)
```
Here you could use that circular buffer, but maybe don't bother yet.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
+        b = ms_coefficients4
+    else
+        #calculating higher order coefficients for implicit Adam Multon method
+        coef = zeros(order+1, order+1)
```
Just use `order` to index into `coef` instead of `k`. i.e. `order==k+1`. Do that too in the main integration loop.

obiajulu replied:
Yeah, this will be easier to read. I'll make the change.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
+                # Assign in correct order for multiplication below
+                #  (a factor depending on j and s) .* (an integral of a polynomial with -(-1:s-1), except -(j-1), as roots)
+                p_int = polyint(poly(diagm(-[-1:j - 2; j:s-1])))
+                coef[s+1, s+1 - j] = ((-1)^j / factorial(j)
```
Have you checked that this gives the right answer for the known coefficients for `k=0..3`?  Maybe double check for the explicit ones too.

obiajulu replied:
Yup, this is giving the correct coefficients for both the implicit and explicit k = 0...3 cases

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
+##Function: ode_imp_ab(F,y0, tspan, order(optional))
+##order is set to 4 by default
+##Adam Bashforth is a fixed step, fixed order, multistep Implicit method
+function ode_imp_ab(F::Function,y0, tspan,order=4 ::Integer)
```
`order=4 ::Integer` does not what you think it should: it checks that 4 is an Integer, then assigns that to `order`. You want `order::Integer=4`.

obiajulu replied:
Great. Taken care of.

mauro3 commented on src/adam_bashforth.jl:
```diff
@@ -1,57 +1,76 @@
-# Explicit Adam-Bashforth solvers
+# Implicit Adam-Bashforth solvers
 ####################################
 # (Hairer & Wanner 1996, Vol I, p.357-358
```
Pages need update.

mauro3 commented:
Cool, this looks good! Above some minor comments.

obiajulu committed:
small tweaks in indexing and coding style

obiajulu committed:
update the reference pages from Hairer et al

obiajulu commented:
@mauro3, your circular buffer looks great. I'm trying to think of the best way of including it in the package, so it could be used with ode_ms and ode_imp_ab if we eventually want to merge upstream. What is more in line with previous ODE convention?

obiajulu committed:
remove commented out 'using Adam Bash'

mauro3 commented:
Just put it into your file and worry about ode_ms later.

obiajulu committed:
remove ode21 from solvers

mauro3 commented:
Well, I'd probably keep the GSoC stuff in their branches and let master of your fork be the same as upstream master.

obiajulu commented:
Yeah, that's what I'm trying to do right now. I'm making this ob/a-b, and trying to get master to be the same as upstream/master
